modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -0,0 +1,331 @@
+            return SupervisorLogs(stage_runs=[])
+        df_trainings = df_triggers[["trigger_id", "first_timestamp", "last_timestamp"]].merge(
+            df_store_models[["id_model", "training_id", "trigger_id"]], on="trigger_id"
+        )
+
+        eval_requests: list[EvalRequest] = []
+        for eval_handler in self.eval_handlers:
+            if eval_handler.config.execution_time != "after_pipeline":
+                continue
+
+            handler_eval_requests = eval_handler.get_eval_requests_after_pipeline(df_trainings=df_trainings)
+            eval_requests += handler_eval_requests
+
+        num_workers = None
+        if self.config.supervisor:
+            num_workers = self.config.supervisor.after_pipeline_evaluation_workers
+
+        logs = self._launch_evaluations_async(
+            eval_requests,
+            parent_log=StageLog(
+                id=PipelineStage.EVALUATE_SINGLE.name,
+                start=datetime.datetime.now(),
+                batch_idx=-1,
+                sample_idx=-1,
+                sample_time=-1,
+                trigger_idx=-1,
+            ),
+            eval_status_queue=eval_status_queue,
+            num_workers=num_workers,
+        )
+        return logs
+
+    # -------------------------------------------------------------------------------- #
+    #                                     Internal                                     #
+    # -------------------------------------------------------------------------------- #
+
+    def _launch_evaluations_async(
+        self,
+        eval_requests: list[EvalRequest],
+        parent_log: StageLog,
+        eval_status_queue: Queue,
+        num_workers: int | None = None,
+    ) -> SupervisorLogs:
+        """Creates a thread pool to launch evaluations asynchronously.
+
+        Args:
+            eval_requests: The evaluation requests to launch.
+            num_workers: The number of workers to use.
+        """
+
+        tasks: list[Future[StageLog]] = []
+        logs = SupervisorLogs()
+
+        def worker_func(eval_req: EvalRequest) -> StageLog:
+            single_log = StageLog(
+                id=PipelineStage.EVALUATE_SINGLE.name,
+                start=datetime.datetime.now(),
+                batch_idx=parent_log.batch_idx,
+                sample_idx=parent_log.sample_idx,
+                sample_time=parent_log.sample_time,
+                trigger_idx=parent_log.trigger_idx,
+            )
+            epoch_nanos_start = current_time_micros()
+            self._single_evaluation(single_log, eval_status_queue, eval_req)
+            single_log.end = datetime.datetime.now()
+            single_log.duration = datetime.timedelta(microseconds=current_time_micros() - epoch_nanos_start)
+            return single_log
+
+        # As we are io bound by the evaluator server, GIL locking isn't a concern, so we can use multithreading.
```

Comment:
I am actually surprised this works, because using gRPC in a multithreaded fashion at least used to be a problem. But I also did not encounter issues so far, so until we fix the evaluator, let's keep this